sdks/python/apache_beam/examples/complete/distribopt/distribopt/distribopt.py

```diff
@@ -0,0 +1,349 @@
+      solution = (mapping_identifier, (greenhouse, optimum))
+      yield pvalue.TaggedOutput('solution', solution)
+      for crop, cost, quantity in zip(crops, costs, quantities):
+        costs = (crop, greenhouse, mapping_identifier, cost * quantity)
+        yield pvalue.TaggedOutput('costs', costs)
+
+  def expand(self, inputs):
+    mappings, quantities = inputs
+    opt = (
+        mappings
+        | 'optimization tasks' >> beam.ParDo(self.CreateOptimizationTasks(), pvalue.AsDict(quantities))
+        | 'optimize' >> beam.ParDo(self.OptimizeProductParameters()).with_outputs('costs', 'solution')
+    )
+    return opt
+
+
+class CreateTransportData(beam.DoFn):
+  """
+  Transform records to pvalues ((crop, greenhouse), transport_cost)
+  """
+
+  def process(self, record):
+    crop = record['crop']
+    for greenhouse, transport_cost in record['transport_costs']:
+      yield ((crop, greenhouse), transport_cost)
+
+
+def add_transport_costs(element, transport, quantities):
+  """
+  Adds the transport cost for the crop to the production cost.
+
+  pvalues are of the form (crop, greenhouse, mapping, cost), the cost only
+  corresponds to the production cost. Return the same format, but including
+  the transport cost.
+  """
+  crop = element[0]
+  cost = element[3]
+  # lookup & compute cost
+  transport_key = element[:2]
+  transport_cost = transport[transport_key] * quantities[crop]
+  return element[:3] + (cost + transport_cost,)
+
+
+def parse_input(line):
+  # Process each line of the input file to a dict representing each crop
+  # and the transport costs
+  columns = line.split(',')
+
+  # Assign each greenhouse a character
+  transport_costs = []
+  for greenhouse, cost in zip(string.ascii_uppercase, columns[2:]):
+    info = (greenhouse, int(cost) if cost else None)
+    transport_costs.append(info)
+
+  return {
+    'crop': columns[0],
+    'quantity': int(columns[1]),
+    'transport_costs': transport_costs
+  }
+
+
+def format_output(element):
+  """
+  Transforms the datastructure (unpack lists introduced by CoGroupByKey)
+  before writing the result to file.
+  """
+  result = element[1]
+  result['cost'] = result['cost'][0]
+  result['production'] = dict(result['production'])
+  result['mapping'] = result['mapping'][0]
+  return result
+
+
+def run(argv=None):
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--input',
+                      dest='input',
+                      required=True,
+                      help='Input description to process.')
+  parser.add_argument('--output',
+                      dest='output',
+                      required=True,
+                      help='Output file to write results to.')
+  known_args, pipeline_args = parser.parse_known_args(argv)
+  pipeline_options = PipelineOptions(pipeline_args)
+  pipeline_options.view_as(SetupOptions).save_main_session = True
+
+  with beam.Pipeline(options=pipeline_options) as pipe:
```

Comment:
Please follow a similar naming to other examples. This would allow readers to recognize common parts easily. For example in this case you could use `p` instead of `pipe`.